sklearn/model_selection/_search.py

```diff
@@ -513,84 +518,132 @@ def _fit(self, X, y, labels, parameter_iterable):
         cv = check_cv(self.cv, y, classifier=is_classifier(estimator))
         self.scorer_ = check_scoring(self.estimator, scoring=self.scoring)
 
-        n_samples = _num_samples(X)
         X, y, labels = indexable(X, y, labels)
-
-        if y is not None:
-            if len(y) != n_samples:
-                raise ValueError('Target variable (y) has a different number '
-                                 'of samples (%i) than data (X: %i samples)'
-                                 % (len(y), n_samples))
         n_splits = cv.get_n_splits(X, y, labels)
-
         if self.verbose > 0 and isinstance(parameter_iterable, Sized):
             n_candidates = len(parameter_iterable)
             print("Fitting {0} folds for each of {1} candidates, totalling"
                   " {2} fits".format(n_splits, n_candidates,
                                      n_candidates * n_splits))
+        candidate_params = list(parameter_iterable)
```

Comment:
I know it's untested and undocumented, but let's leave the code able to handle a `parameter_iterable` that can be passed through once, and which may involve sophisticated computation. Let `return_parameters=True` and pull the `candidate_params` out in the `zip(*out)` unpacking below.